docs/TypeInference.md

```diff
@@ -0,0 +1,218 @@
+In this system record kinds have a set of properties along with their types.
+They also contain two sets of labels.
+One set specifies the properties that a record **must** have, the other specifies the properties it **may** have.
+While this allows for polymorphic record access as well as polymorphic record literals, it lacks support for record extensions.
+The following example cannot be typed in this system.
+
+```
+r = {a: 1, b: 2}
+s = {r with c: 3} // s is a record with properties a, b, and c
+```
+
+In order to support examples like the one above, I propose we replace the current record kind with one that allows for extension such as the one in [this paper](https://www.microsoft.com/en-us/research/wp-content/uploads/2016/02/scopedlabels.pdf).
+Here the syntax of record types more closely aligns with the syntax of the values themselves.
+In this system records are constructed via extension, starting from the empty record `{}`.
+
+For example the record literal `{a:1, b:2}` is just shorthand for `{a:1 | {b:2 | {} }}` whose type is given by `{a:int | {b:int | {} }}`.
+
+More generally, the polymorphic type of the extension operation `{_ with l:_}` is given by:
+
+```
+forall ['a, 'r] where 'r: Rec ('a, 'r) -> {l: 'a | 'r}
+```
+
+This type is read as, "for all type variables _'a_ and _'r_ where _'r_ is a record kind, we have a function type taking _'a_ and _'r_ as input and returning a record type extended from _'r_ by adding a field _l_ of type _'a_".
+Selection and restriction operations are defined similarly, and update and rename can be expressed using these basic operations.
+
+Unification of record types is particularly elegant and can be expressed using the following 3 rules:
+
+1. `{a: 't | 'r} = {a: 'u | 's} => 't = 'u and 'r = 's`
+2. `{a: 't | 'r} = {b: 'u | 's} => {a: 't | 'v} = 's and {b: 'u | 'v} = 'r`
+3. `{a: 't | 'r} = {b: 'u | 'r} => error`
+
+Note here `=` is synonymous for unify.
+
+### Implications
+
+The Flux type system extends Hindley-Milner with a _kind_ system.
+Currently records are modeled as _kinds_ in this system.
+Extensible records however fall entirely within the _type_ system, not the _kind_ system.
+Describing records this way allows us to break the mutual recursion between type unification and kind unification.
+Proving the correctness of type inference now follows directly from Hindley-Milner.
+
+### Caveats
+
+* Nulls
+
+    Extensible records do not support accessing non-existent fields on an record.
+    Such operations will result in a unification error.
+    The Flux language specification states that accessing a non-existent field on a record will return an untyped null value.
+    The spec will need to be updated accordingly as this will no longer be the case.
+
+    Null values will still be possible at runtime but they must be typed nulls.
+    The concept of an untyped null will need to be removed from the spec.
```

Comment:
I don't follow this bit, if we know that that field doesn't exist on the record then we know it will be a null when we access it. So we know it would have any other type but null. So where does the untyped vs typed nulls part come in?

Reply:
Unifying extensible records can only be well-defined if we error when a non-existent field is accessed. An example might help clarify.

Say I have a function that takes in a record that needs an `a` field of type _int_.
```
typeof f = forall ['t] where 't:Rec (r: {a:int | 't}) -> int
```
Then we call `f` with a record that does not have `a`:
```
f(r: {b: 3})
```
This results in the following unification implications:
```
{a:int | 't} = {b:int | {} }
    => 't = {b:int | 's} and {} = {a:int | 's}
```
How do we unify `{}` with `{a: int | 's}`?

I guess we could say that `'s = {}`, but what do we do with `a: int`? We've already inferred that `a` is an _int_. _int_ is a concrete monotype and we can't just switch it to be _null_. This has to be a unification error.

At runtime however, in the case of row functions, this won't happen. This is because if `f` is a row function defined in `map` or `filter`, it's never called in the language itself. It's execution is implied by the internal implementation of `map` or `filter`. Hence null values can exist as part of an integer column, the type system just sees them as integers and the interpreter/execution engine operates on them according to their semantics as defined in the spec.